campaignion_layout/src/Lookup.php

```diff
@@ -70,10 +70,12 @@ public function getTheme() {
    * @return string|null
    *   The name of the first enabled layout of the currently active theme.
    */
-  public function getLayout() {
-    foreach ($this->iterateItems() as $item) {
-      if (($theme = $this->themes->getTheme($item['theme'])) && $theme->isActive() && ($layout = $theme->getLayout($item['layout']))) {
-        return $layout;
+  public function getLayout(string $active_theme_name) {
+    if ($theme = $this->themes->getTheme($active_theme_name)) {
+      foreach ($this->iterateItems() as $item) {
+        if ($active_theme_name == $item['theme'] && ($layout = $theme->getLayout($item['layout']))) {
```

Comment:
Just wondering, could this further be simplified by merging `getLayout()` and `getTheme()` into a `getItem()` function (as far as I can see they are both only called 1x)? Is there a chance that `$active_theme_name` is different to the one returned by `getTheme()`?

Reply:
Hm, this is tricky as `Themes::getTheme()` and `Theme::getLayout()` are implemented on different classes and work with different data. How do you think about the new `Theme::getLayoutFromItems()`?

Reply:
Oh no, I was only referring to the `Lookup` class!

```php
  // E.g. `Lookup::getTheme()` could return the item instead of the theme
  public function getItem() {
    foreach ($this->iterateItems() as $item) {
      if (($theme = $this->themes->getTheme($item['theme'])) && $theme->hasFeatureEnabled()) {
        return $item;
      }
    }
  }

// Then somewhere else
$theme = $lookup->getItem()['theme'];
$layout = $lookup->getItem()['layout'];
```

But `Theme::getLayoutFromItems()` is also an improvement to the previous code :)

Reply:
That would still not work:

- For the theme we choose the first item with a feature-enabled theme.
- For the layout we choose the first item of the currently active theme which contains a layout that’s enabled for this theme.

Your code would use the same criteria for both to choose the item.

Reply:
That’s was what I was wondering in the first comment, *is* that a difference? Shouldn’t the first selected theme be the theme that is used?

Reply:
@katzmo If we assume nothing else interferes with theme selection: yes. Even then it’s still not as simple:

```php
$items[] = ['theme' => 'a', 'layout' => 'not available'];
$items[] = ['theme' => 'b', 'layout' => 'available'];
$items[] = ['theme' => 'a', 'layout' => 'available'];
```

This should choose the third item, not the first. (Assuming both themes have the feature enabled.)